pkg/shell/hosts.jsx

```diff
@@ -0,0 +1,232 @@
+        return ReactDOM.createPortal(children, this.el);
+    }
+}
+
+function HostLine(props) {
+    const host = props.host;
+    const user = props.user;
+    return (
+        <>
+            <span className="username">{user }</span>
+            {user &&
+            <span className="at">@</span>
+            }
+            <span className="hostname">{host}</span>
+        </>
+    );
+}
+
+export class cHosts extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            opened: false,
+            editing: false,
+            current_user: "",
+        };
+
+        this.toggleMenu = this.toggleMenu.bind(this);
+        this.filterHosts = this.filterHosts.bind(this);
+        this.onAddNewHost = this.onAddNewHost.bind(this);
+        this.onEditHosts = this.onEditHosts.bind(this);
+        this.onHostEdit = this.onHostEdit.bind(this);
+        this.onRemove = this.onRemove.bind(this);
+
+        this.machines = machines.instance();
+        this.mdialogs = new_machine_dialog_manager(this.machines);
+    }
+
+    componentDidMount() {
+        cockpit.user().then(user => {
+            this.setState({ current_user: user.name || "" });
+        });
+    }
+
+    toggleMenu() {
+        hosts_sel.hidden = this.state.opened;
+        this.setState(s => { return { opened: !s.opened } });
+    }
+
+    onAddNewHost() {
+        this.mdialogs.render_dialog("add-machine", "hosts_setup_server_dialog");
+    }
+
+    onHostEdit(event, host) {
+        const dlg = $("#edit-host-dialog");
+
+        const machine = this.machines.lookup(host.key);
+        if (!machine)
+            return;
+
+        const can_change_user = machine.address != "localhost";
+        const name = document.getElementById("edit-host-name");
+        name.disabled = machine.state == "failed";
+        name.value = machine.label;
+
+        document.getElementById("edit-host-user-row").hidden = !machines.allow_connection_string;
+        const user = document.getElementById("edit-host-user");
+        if (machines.allow_connection_string) {
+            user.placeholder = this.state.current_user;
+            user.disabled = !can_change_user;
+            user.value = machine.user || "";
+            $("#edit-host-dialog a[data-content]").popover();
+        }
+
+        this.mdialogs.render_color_picker("#edit-host-colorpicker", machine.address);
+
+        document.getElementById("edit-host-sync-users").addEventListener("click", e => {
+            dlg.modal('hide');
+            this.mdialogs.render_dialog("sync-users",
+                                        "hosts_setup_server_dialog",
+                                        machine.address);
+        });
+
+        // Remove all existing listeners so we don't change it multiple times
+        const orig = document.getElementById("edit-host-apply");
+        var copy = orig.cloneNode(true);
+        orig.parentNode.replaceChild(copy, orig);
```

Comment:
Oh wow -- are we missing a `removeEventListener()` (or the corresponding jquery) somewhere? So cloneNode() will copy the DOM but not attached listeners? But doesn't that mean that the original node is still around somewhere and will invoke event listeners?

Reply:
This is hack I did not get back to (kinda as I am not aware of a better solution).
I set up listener right underneath this 3 lines. But when the dialog is closed (either by 'Cancel' or by applying it), the listeners stay around. I would need to remove events in this two cases, but I cannot just do 'removeEvent' as I don't have the function. 

cloneNode does not copy listeners. And the original node exists in `shell/index.html`

Reply:
Seems ok, at  least the comment explains what's going on. I was just curious. This kind of hack should go away with a full Reactification. :)